crates/core/src/execution_context.rs

```diff
@@ -1,17 +1,127 @@
+    pub fn new(table_id: TableId, table_name: String) -> Self {
+        Self {
+            table_id,
+            index_seeks: 0,
+            keys_scanned: 0,
+            rows_fetched: 0,
+            cache_table_name: table_name,
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Metrics(Vec<BufferMetric>);
+impl Metrics {
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn inc_by<F: FnOnce() -> String>(&mut self, table_id: TableId, ty: MetricType, val: u64, get_table_name: F) {
+        if let Some(metric) = self.0.iter_mut().find(|x| x.table_id == table_id) {
+            metric.inc_by(ty, val);
+        } else {
+            let table_name = get_table_name();
+            let mut metric = BufferMetric::new(table_id, table_name);
+            metric.inc_by(ty, val);
+            self.0.push(metric);
+        }
+    }
+
+    pub fn table_exists(&self, table_id: TableId) -> bool {
+        self.0.iter().any(|x| x.table_id == table_id)
+    }
+
+    fn flush(&mut self, workload: &WorkloadType, database: &Address, reducer: &str) {
+        self.0.iter().for_each(|metric| {
+            DB_METRICS
+                .rdb_num_index_seeks
+                .with_label_values(
+                    workload,
+                    database,
+                    reducer,
+                    &metric.table_id.0,
+                    &metric.cache_table_name,
+                )
+                .inc_by(metric.index_seeks);
+
+            DB_METRICS
+                .rdb_num_keys_scanned
+                .with_label_values(
+                    workload,
+                    database,
+                    reducer,
+                    &metric.table_id.0,
+                    &metric.cache_table_name,
+                )
+                .inc_by(metric.keys_scanned);
+
+            DB_METRICS
+                .rdb_num_rows_fetched
+                .with_label_values(
+                    workload,
+                    database,
+                    reducer,
+                    &metric.table_id.0,
+                    &metric.cache_table_name,
+                )
+                .inc_by(metric.keys_scanned);
+        });
+    }
+}
 
 /// Represents the context under which a database runtime method is executed.
 /// In particular it provides details about the currently executing txn to runtime operations.
 /// More generally it acts as a container for information that database operations may require to function correctly.
-#[derive(Default)]
-pub struct ExecutionContext<'a> {
+#[derive(Default, Clone)]
+pub struct ExecutionContext {
     /// The database on which a transaction is being executed.
     database: Address,
     /// The reducer from which the current transaction originated.
-    reducer: Option<&'a str>,
+    reducer: Option<String>,
     /// The type of workload that is being executed.
     workload: WorkloadType,
+    /// The Metrics to be reported for this transaction.
+    pub metrics: Arc<RwLock<Metrics>>,
```

Comment:
Will there ever be concurrent access that this needs to be behind a mutex?

Reply:
Yes, I can think of subscriptions where multiple threads may use same ctx.